components/butterfly/src/network.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The Butterfly network abstraction.
+//!
+//! The abstraction provides sockets for sending SWIM and gossip
+//! messages.
+
+use std::cell::UnsafeCell;
+use std::error::Error as StdError;
+use std::fmt::{Debug, Error as FmtError, Formatter};
+use std::marker::{Send, Sized};
+use std::net::{IpAddr, Ipv4Addr, SocketAddr, UdpSocket};
+use std::result::Result as StdResult;
+use std::time::Duration;
+
+use error::{Result, Error};
+
+use zmq;
+
+use ServerContext;
+
+/// A trait for types that can fail when cloning.
+pub trait TryClone: Sized {
+    fn try_clone(&self) -> Result<Self>;
+}
+
+// TODO(krnowak): See a TODO about Debug for Network trait below.
+/// A trait for types used for communicating with SWIM messages.
+pub trait SwimSocket: TryClone + Send + Debug {
+    /// Sends given bytes to the given address. The returned value
+    /// holds a number of bytes sent.
+    fn send_to(&self, buf: &[u8], addr: SocketAddr) -> Result<usize>;
+    /// Receive bytes from the socket. The returned value holds the
+    /// size and an address from where the bytes came.
+    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, SocketAddr)>;
+}
+
+/// A trait for types used for sending gossip messages.
+pub trait GossipPushSocket {
+    /// Send bytes being a gossip message.
```

Comment:
I don't understand this sentence.

Reply:
Abracadabra, hocus pocus, what's here not to understand? ;)

Right, it's terrible. Will reword it.